### DISCUSSION TITLE
Building with .net framework 4.8

### DISCUSSION THREAD

**Discussion by Leon-Cantin** [EliotVU/Unreal-Library#80] [Q&A]
Hello, I am trying to decompile some files for Endwar which is based on UE3 and so I tried to open the .upk but I am getting some errors so I tried building the projet and debugging locally with vscode to see if I could figure out what is going on and maybe add support for the game if I figure it out.
I tried building with vscode at first but it seems to use dotnet cli which is not the way the .net sdk framework 4.8 works. It uses MSBuild (thanks MS for making this so confusing)
So I finally figured that running with the command
`C:\Windows\Microsoft.NET\Framework64\v4.0.30319\msbuild "E:\Development\UDK\UE-Explorer\UE Explorer.sln"`
would allow to build with your specific .net version but I am getting an error from MSBuild that all of the .csproj are not supported by MSBuild. Does this ring any bell?

I am also getting errors like:
`E:\Development\UDK\UE-Explorer\UELib\src\Eliot.UELib.csproj(1,1): error MSB4041: L'espace de noms XML par défaut du projet doit être l'espace de noms XML MSBuild. Si le projet a été créé au
 format MSBuild 2003, ajoutez xmlns="http://schemas.microsoft.com/developer/msbuild/2003" à l'élément <Project>. S'il a été créé dans des formats anciens comme 1.0 ou 1.2, convertissez-le a
u format MSBuild 2003.`
which seem linked to the issue. Saying something like the default XML project namespace must be the MSBuild namespace. I tried adding the xml element above but I am getting more errors.

Maybe you ran into such an issue before?

Thanks

**Comment by EliotVU** [comment]
That's unfortunate :( 

I gave it a try myself, I ran into a few issues but they were easy to fix, however I couldn't reproduce your build error.

Have you tried this command: `dotnet build UEExplorer.csproj /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary`, I used the *C# Dev Kit* extension.

**Reply by Leon-Cantin** [reply]
Thanks for the very quick answer! I did at first but I found out that dotnet is the CLI which is not included in the download link you gave for 4.8 though I had it already from another version.

dotnet seems to require the SDK installation type which installs in C:\program files\dotnet\sdk which is not how the 4.8 works for some wierd reasons. It even seems to be the only one listed in the "framework" style which instead only installs in C:\windows\Microsoft.NET\Framework64. It is incompatible with "dotnet" on my end. Perhaps you got yourself a different download?

**Reply by EliotVU** [reply]
What does cmd echo if you input `where dotnet`? For me it echos `C:\Program Files\dotnet\dotnet.exe`

**Reply by EliotVU** [reply]
I've had a similar issue before where I had multiple installations of dotnet for some reason, and thus dotnet was executing another installed version.

**Reply by Leon-Cantin** [reply]
`C:\Program Files\dotnet\dotnet.exe` same as you. But it does not seems to be the style of download available on the 4.8 page. I tried uninstalling and reinstalling everything multiple times to figure out what installed what and it seems that this link runs in a different style. It does not work with dotnet CLI.

if you look at this page and the URL [https://dotnet.microsoft.com/en-us/download/dotnet-framework](https://dotnet.microsoft.com/en-us/download/dotnet-framework), you'll notice 3.5 and 4.6 to 4.8 and the mention of "framework".
If you go to this page instead [https://dotnet.microsoft.com/en-us/download/dotnet](https://dotnet.microsoft.com/en-us/download/dotnet), you'll notice that there is no mention of "framework" and you have every version but 4.X.

Found this link: https://stackoverflow.com/questions/69795081/how-can-i-build-a-net-framework-4-8-application-using-the-dotnet-cli
 which seems to confirm my suspissions. I will have to keep digging.

If you do `dotnet --list-sdks` what do you get?

**Reply by EliotVU** [reply]
I get:

2.1.500 [C:\Program Files\dotnet\sdk]
3.1.413 [C:\Program Files\dotnet\sdk]
6.0.202 [C:\Program Files\dotnet\sdk]
6.0.301 [C:\Program Files\dotnet\sdk]
7.0.407 [C:\Program Files\dotnet\sdk]
8.0.202 [C:\Program Files\dotnet\sdk]

**Reply by Leon-Cantin** [reply]
exactly, there is no 4.8! It must be building using something else.

If you run using the C# dev kit in VS Code, you might be able to glean which version it is trying to build with. Here's an example of my output,, but I can only see the version in the path used in errors:
dotnet build e:\Development\UDK\UE-Explorer\UE Explorer\UEExplorer.csproj /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary 
Version MSBuild 17.9.8+b34f75857 pour .NET
  Identification des projets à restaurer...
  Restauration effectuée de e:\Development\UDK\UE-Explorer\Eliot.Utilities\Eliot.Utilities.csproj (en 107 ms).
  Restauration effectuée de e:\Development\UDK\UE-Explorer\UELib\src\Eliot.UELib.csproj (en 265 ms).
  Restauration effectuée de e:\Development\UDK\UE-Explorer\UE Explorer\UEExplorer.csproj (en 290 ms).
C:\Program Files\dotnet\sdk\8.0.204\Microsoft.CSharp.CurrentVersion.targets(129,9): warning MSB3884: le fichier d'ensemble de règles "GlobalizationRules.ruleset" est introuvable. [e:\Development\UDK\UE-Explorer\UELib\src\Eliot.UELib.csproj]
  Eliot.Utilities -> e:\Development\UDK\UE-Explorer\Eliot.Utilities\bin\Debug\net48\Eliot.Utilities.dll
  Eliot.UELib -> e:\Development\UDK\UE-Explorer\UELib\src\bin\Debug\net48\Eliot.UELib.dll
C:\Program Files\dotnet\sdk\8.0.204\Microsoft.Common.CurrentVersion.targets(3386,5): error MSB3823: les ressources qui ne sont pas des chaînes nécessitent la propriété GenerateResourceUsePreserializedResources avec la valeur true. [e:\Development\UDK\UE-Explorer\UE Explorer\UEExplorer.csproj]

**Reply by EliotVU** [reply]
```
dotnet build C:\Projecten\UEExplorerHotFix\UE Explorer.sln /target:UEExplorer /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary 
MSBuild version 17.9.6+a4ecab324 for .NET
```

An older build 🤔

**Reply by EliotVU** [reply]
Ah, and `C:\Program Files\dotnet\sdk\8.0.202\Microsoft.CSharp.CurrentVersion.targets(129,9): warning MSB3884: Could not find rule set file "GlobalizationRules.ruleset". [C:\Projecten\UEExplorerHotFix\UELib\src\Eliot.UELib.csproj]`

**Reply by Leon-Cantin** [reply]
So it seems to be building with a MSBuild version very close to mine and the SDK is too. I downloaded the latest version 8 yesterday to get dotnet cli.

So your version is building just fine with VS Code even though it's using the 8.x SDK?

**Reply by EliotVU** [reply]
Yep, I tried both with the cmd "dotnet" as well as the extension's integrated "Build" button, to see if that could make a difference, both worked.

**Reply by EliotVU** [reply]
When developing, I generally don't use VS Code for C# projects, I use Visual Studio instead (better integration), perhaps VS is installing additional utilities?

**Reply by Leon-Cantin** [reply]
I have a version of VS 2017. When opening the project with VS it told me I could download another extension, will give that a try.

**Reply by Leon-Cantin** [reply]
Still not working, will keep on investigating.

**Reply by Leon-Cantin** [reply]
oh wow I think I got to work. It compiled and ran. I added `<GenerateResourceUsePreserializedResources>true</GenerateResourceUsePreserializedResources>` to the first PropertyGroup of both csproj (although I think you only need need for the lib one).
I also ran `dotnet add package System.Resources.Extensions` inside both projects which adds a reference.

**Reply by Leon-Cantin** [reply]
Well now I am lost. after a one or two run it crashes on boot with an exception about resources. Probably because of what I added to fix the first issue. Some warnings in VS Code also become errors in VS.

Do you still plan on updating it to  a more recent framework?

**Reply by Leon-Cantin** [reply]
Ok, another update. I got it to be stable.

I have `<GenerateResourceUsePreserializedResources>true</GenerateResourceUsePreserializedResources>` on the explorer project and `dotnet add package System.Resources.Extensions` ran on the lib project.

Now I gotta figure out why the exception just show a window to report them instead of being caught by the debugger.

**Reply by EliotVU** [reply]
I will update the project .NET to the latest version probably with release 2.0

**Reply by EliotVU** [reply]
"Now I gotta figure out why the exception just show a window to report them instead of being caught by the debugger." 

This usually happens if it is caught before any Program.Main code is executed.

**Reply by Leon-Cantin** [reply]
I figured, using break points I managed to find the spot. I wonder if there's a way to always catch them. Also building with VsCode works but the break points don't and a building with VS does not work but the break points work. At least i got a working workflow.

The program crashes somewhere in the early bytes where it tries to get the size for an array which ends up being huge. It's a very small file so it's obviously wrong. Do you have any tips on how to figure out the layout of a UPK?

**Reply by EliotVU** [reply]
Nice :)

I'd recommend Ghidra, analyze the respective .exe and look for a "UTF-16" string `binaryformat` it should be referred by one function, which should be the one that's deserializing the UPK's header.

**Reply by Leon-Cantin** [reply]
Well, I didn't think I'd have to take out the big guns so soon. I was looking forward to using Ghidra. Thanks a lot for the tip.

**Reply by Leon-Cantin** [reply]
![image](https://github.com/EliotVU/Unreal-Library/assets/755263/8555af25-a23d-4ae2-a163-908528d55f14)

I think I got my bearing in the software but it seems tehre is no references to this adress.

**Reply by EliotVU** [reply]
That's unfortunate, I have ran into that issue with many older UE3 based games :/

Nonetheless, I gave it a go myself and got this:
![image](https://github.com/EliotVU/Unreal-Library/assets/808593/352fc244-8530-461d-9a3f-f6c8f153c529)

**Reply by EliotVU** [reply]
Which lead me to this:

![image](https://github.com/EliotVU/Unreal-Library/assets/808593/16620872-d4e2-4d6e-b902-81ec82b31b43)

and the function we want is all the way below the error:
![image](https://github.com/EliotVU/Unreal-Library/assets/808593/6999d8d0-4296-4317-8e17-c8e11ba73bfb)

**Reply by Leon-Cantin** [reply]
Wow, thanks a lot for digging into it. How did you find that other refence to the string? It might be me not knowing the software very well. I went to these adresses but I can't see those functions.

**Reply by EliotVU** [reply]
Try run One-Shot->ASCII Strings

**Reply by Leon-Cantin** [reply]
Still does not show up when doing a memory search. When I run the auto-analysis with the default parameters. I get these errors:
![image](https://github.com/EliotVU/Unreal-Library/assets/755263/52760695-4939-4960-8186-8fb9659de36c)

I don't know if that has anything to do with it.

**Reply by Leon-Cantin** [reply]
It seems your analyser managed to decompile and mark a few functions which mine did not, like it stopped somewhere.

**Reply by Leon-Cantin** [reply]
I found the logs and after running an auto-analyse I get these logs:

```
2024-05-12 21:37:45 INFO  (LoggingInitialization) Using log config file: jar:file:/E:/Development/ghidra_11.0.3_PUBLIC/Ghidra/Framework/Generic/lib/Generic.jar!/generic.log4j.xml  
2024-05-12 21:37:45 INFO  (LoggingInitialization) Using log file: C:\Users\Big Boss\.ghidra\.ghidra_11.0.3_PUBLIC\application.log  
2024-05-12 21:37:45 INFO  (Preferences) Loading user preferences: C:\Users\Big Boss\.ghidra\.ghidra_11.0.3_PUBLIC\preferences  
2024-05-12 21:37:45 INFO  (ClassSearcher) Searching for classes...  
2024-05-12 21:37:47 INFO  (ClassSearcher) Class search complete (1592 ms)  
2024-05-12 21:37:47 INFO  (SSLContextInitializer) Initializing SSL Context  
2024-05-12 21:37:47 INFO  (SecureRandomFactory) Initializing Random Number Generator...  
2024-05-12 21:37:47 INFO  (SecureRandomFactory) Random Number Generator initialization complete: SHA1PRNG  
2024-05-12 21:37:47 INFO  (ApplicationTrustManagerFactory) Trust manager disabled, cacerts have not been set  
2024-05-12 21:37:47 INFO  (GhidraRun) User BigBoss started Ghidra.  
2024-05-12 21:37:48 DEBUG (RecoverySnapshotMgrPlugin) Recovery snapshot timer set to 5 minute(s)  
2024-05-12 21:37:49 INFO  (DefaultProject) Opening project: E:\Development\endwar\endwar2  
2024-05-12 21:37:51 INFO  (PackedDatabaseCache) Packed database cache: C:\Users\Big Boss\AppData\Local\Ghidra\packed-db-cache  
2024-05-12 21:37:52 DEBUG (PackedDatabaseCache) Using cached packed database: E:\Development\ghidra_11.0.3_PUBLIC\Ghidra\Features\Base\data\typeinfo\win32\windows_vs12_32.gdt  
2024-05-12 21:38:04 DEBUG (PackedDatabaseCache) Using cached packed database: E:\Development\ghidra_11.0.3_PUBLIC\Ghidra\Features\Base\data\typeinfo\win32\windows_vs12_32.gdt  
2024-05-12 21:38:42 DEBUG (ToolTaskManager) Background processing started...  
2024-05-12 21:38:42 DEBUG (ToolTaskManager) Exec Task Auto Analysis  
2024-05-12 21:38:42 DEBUG (SymbolServerService) SymbolServerService: querying Program's Import Location - E:\Jeux\SteamLibrary\steamapps\common\Tom Clancys Endwar\Binaries for PCLaunch-OfGame_vc8.pdb, aa8bd605-381c-4acb-8332-cbfc3a0bd18a, 4, 0, ???  
2024-05-12 21:38:42 DEBUG (SymbolServerService) SymbolServerService: got 0 results from Program's Import Location - E:\Jeux\SteamLibrary\steamapps\common\Tom Clancys Endwar\Binaries  
2024-05-12 21:38:42 DEBUG (SymbolServerService) SymbolServerService: found 0 matches  
2024-05-12 21:38:42 INFO  (PdbUniversalAnalyzer) Skipping PDB processing: failed to locate PDB file in configured locations  
2024-05-12 21:38:42 INFO  (PdbUniversalAnalyzer) You may set the PDB "Symbol Server Config"
 using "Edit->Symbol Server Config" prior to analysis.
It is important that a PDB is used during initial analysis 
if available.  
2024-05-12 21:38:44 ERROR (PngDataType) Invalid PNG data at 11fddbf0  
2024-05-12 21:38:49 WARN  (DecompileCallback) Decompiling 11807120, pcode error at 118071c6: Unable to resolve constructor at 118071c6  
2024-05-12 21:39:28 INFO  (ApplyDataArchiveAnalyzer) Applied data type archive: windows_vs12_32  
```

Your auto-analyse seems to catch much more stuff for some reasons. Mine stops after like a minute which seems way too quick. It seems to stop too early.

**Reply by EliotVU** [reply]
I'm using this plugin: https://github.com/CERTCC/kaiju but I am not sure if that is of any help with this.

**Reply by EliotVU** [reply]
Also, my configuration: 
![image](https://github.com/EliotVU/Unreal-Library/assets/808593/db8db2bb-1ed2-4c9e-89c3-e4ef30942c94)

**Reply by Leon-Cantin** [reply]
Damn, still not working. It's not only the string that is the issue, even if I go at the adress you mentioned near the start of this conversation, Ghidra has not dissassembled any function there.
I am just using Ghidra 11.0.3 out of the box and running the auto analysis on the latest steam version of endwar. The MD5 hash for the exec is "c44d72cc29fec23ec2422faa12b8368e".

Are you doing anything special when start the project?

**Reply by Leon-Cantin** [reply]
I left Kaiju runnign for a while where it says "Analyzing gaps (Iteration N)" at the bottom right. It's now at iteration 233 but it does not seems to be doing anything. How long does your analysis takes to get to such a nice level of disassembly? From what I have seen online, Ghidra will not disassemble so much by default. You have to do something like selecting a few adresses and pressing D.

**Reply by Leon-Cantin** [reply]
What are the options you are using when you import the .exe? Anything non default?

**Reply by EliotVU** [reply]
All default :/

**Reply by EliotVU** [reply]
hash of my exe file: c049d7933c47d5683117a5e152443711

**Reply by Leon-Cantin** [reply]
That's funny, the MD5 hash is different. I am using the steam version 1.0.0.2.

All you do is load with default, run the auto-analysis and Ghidra ends up finding all of the functions and opcodes? Mine is all unknown opcode unless I select everything and hit D, but I am not sure that dissassembly is correct and it takes forever.

**Reply by EliotVU** [reply]
Some compiled .exes are just wonky with Ghidra :/ Even for me this one is missing the usual strings like "Bad Expr"

btw: I added support (mostly) for EndWar just now, see https://github.com/EliotVU/Unreal-Library/commits/feature/Feature-endwar/

**Reply by Leon-Cantin** [reply]
Wow, thanks a lot for this, I didn't expect you to go this far. This should help us in our endeavour. Can I ask you where you got your EndWar exe?

**Reply by Leon-Cantin** [reply]
I got the lib on the endwar feature branch but I am still getting the same error after recompiling. Is there anything I need to activate?

**Reply by Leon-Cantin** [reply]
I installed Ghidra on a Windows 10 VM and a Kali VM then installed Ghidra on both and decompiled endwar.exe. Every single one gave me the same result. I wonder if it's something special with the steam version ...

**Reply by Leon-Cantin** [reply]
I decided to buy the Ubisoft Connect version and for some reason that one is different from the steam version and from your version with MD5:00632513640A4EE2F45A16DD94E89203.

BUT it does decompile nicely and I found the functions you mentionned at the start. I can now follow along if there's anything left to do.
